### ISSUE TITLE
MA0070 - Obsolete attributes should include explanations is handled by CA1041

### ISSUE THREAD

**Issue by Meir017** [meziantou/Meziantou.Analyzer#853]
- **Rule**: [MA0070](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0070.md) - Obsolete attributes should include explanations
- **Built-in Equivalent**: [CA1041](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1041) - Provide ObsoleteAttribute message

## Analysis

**MA0070** and **CA1041** serve the same purpose: ensuring that `ObsoleteAttribute` includes meaningful explanatory messages to help developers understand why an API is obsolete and what alternatives they should use.

### MA0070 Coverage
- Detects ObsoleteAttribute without explanation messages
- Promotes clear deprecation communication
- Ensures proper API deprecation practices

### CA1041 Coverage  
- Detects ObsoleteAttribute that should provide messages
- Official Microsoft rule for proper API deprecation
- Ensures deprecation messages are informative
- Part of the .NET design guidelines

## Code Examples

**Problematic code detected by both rules:**
```csharp
// Both MA0070 and CA1041 will flag this
public class ApiClass
{
    [Obsolete] // Missing explanation message!
    public void OldMethod()
    {
        // Implementation
    }
    
    [Obsolete("")] // Empty message!
    public string OldProperty { get; set; }
    
    [Obsolete(null)] // Null message!
    public event EventHandler OldEvent;
}
```

**Recommended fix:**
```csharp
// Both rules suggest providing meaningful messages
public class ApiClass
{
    [Obsolete("Use NewMethod instead. This method will be removed in v2.0.")]
    public void OldMethod()
    {
        // Implementation
    }
    
    [Obsolete("Use NewProperty instead. This property lacks validation and will be removed in the next major version.")]
    public string OldProperty { get; set; }
    
    [Obsolete("Use NewEvent instead. This event has been replaced with a more specific event type.", error: true)]
    public event EventHandler OldEvent;
}
```

## Benefits of Good Deprecation Messages

Proper ObsoleteAttribute messages provide:
- **Clear migration guidance** - what to use instead
- **Timeline information** - when removal will occur
- **Reason for deprecation** - why the API is being removed
- **Better developer experience** - reduces confusion and support requests

## Recommendation

**CA1041** is the official Microsoft analyzer rule that provides **equivalent functionality** to MA0070. Since both rules target the same API deprecation best practice and CA1041 is the standard rule for obsolete attributes, **MA0070 should be deprecated** in favor of CA1041.

## Migration Path

1. Enable **CA1041** in your analyzer configuration:
   ```xml
   <WarningsAsErrors Include="CA1041" />
   ```
2. Disable **MA0070**:
   ```xml
   <WarningsNotAsErrors Include="MA0070" />
   ```
3. Update any rule suppressions from MA0070 to CA1041

## Related

- Parent issue: #835 - Consider deprecating rules that are now covered by built-in .NET analyzers

**Comment by Meir017** [comment]
## Analysis: Equivalent Rules - Recommend Deprecation ✅

**MA0070 vs CA1041** - Both ensure `ObsoleteAttribute` includes meaningful explanatory messages.

### Recommendation: Deprecate MA0070 ✅

**CA1041** is the official Microsoft rule for proper API deprecation. **MA0070 should be deprecated** in favor of CA1041.

---
*Part of #835 deprecation initiative.*

**Comment by Meir017** [comment]
## Analysis Complete ✅

**Issue 853 was missing from the comparison file** and **the rules are equivalent**.

### Findings:
- **MA0070** and **CA1041** serve identical purposes - ensuring `ObsoleteAttribute` includes meaningful explanatory messages
- Both detect the same violations (missing, empty, or null messages in `[Obsolete]` attributes)
- Both promote the same best practice for API deprecation

### Resolution:
- ✅ **Added to comparison file** - CA1041 ↔ MA0070 mapping added to the "Equivalent rules" section
- ✅ **Rules are equivalent** - CA1041 is the official Microsoft analyzer rule that covers the same functionality as MA0070

Since CA1041 is the standard .NET analyzer rule for this scenario, **MA0070 could be considered for deprecation** in favor of the official Microsoft rule as part of the broader initiative in #835.

**Closing this issue** as the comparison has been documented and the equivalence confirmed.
